### DISCUSSION TITLE
Thoughts on the UX

### DISCUSSION THREAD

**Discussion by quazar-omega** [saber-notes/saber#163] [Ideas]
Following up from #117 I'll share some ideas here:
I feel like the UX should display that more clearly, right now it seems a little confusing, like the notes would be data that belongs to the app (like an object), a user would expect that, since there is no interaction with the system file picker.

**~~(Very dangerous)~~ thinking out loud ahead so maybe don't consider it for your own sanity**
~~I wonder if the whole paradigm would have to be changed following this line of thought, if~~ we could make it so that by default it saves to a directory visible to the user like  `<platform-specific-path>/Documents/Saber` which would also work towards #24
This could also open the possibility to optionally
- "save as" somewhere else, invoking the platform file save dialog (..or not since, in my experience, I saw that the state of cross platform file pickers on Flutter isn't the best, maybe things have changed now)
- "open with" Saber when you click on those files, but that comes with the overhead of needing a database that keeps track of where are the files you opened to display the list within the app, unless you decide to only do that in the Saber directory of course.
That would be a sound decision as well imo.

<sup>For a while pretend that I didn't write any of this</sup>, enjoy your holidays!

**Comment by adil192** [comment]
> if we could make it so that by default it saves to a directory visible to the user like  `<platform-specific-path>/Documents/Saber` which would also work towards #24 

That's actually how it currently works! On desktop, we store notes in "Documents/Saber", but on Android/iOS, the "documents" folder is hidden from the user. See [`getApplicationDocumentsDirectory()`](https://pub.dev/documentation/path_provider/latest/path_provider/getApplicationDocumentsDirectory.html). 


>   - "save as" somewhere else, invoking the platform file save dialog (..or not since, in my experience, I saw that the state of cross platform file pickers on Flutter isn't the best, maybe things have changed now)

I may add this in the future (maybe restricted to desktop platforms?). I'm already using [file_picker](https://pub.dev/packages/file_picker) for picking images on desktops and it seems pretty solid (Android and iOS use [image_picker](https://pub.dev/packages/image_picker) to avoid needing unnecessary permissions).
I'd like to keep this app mostly as a self contained tool (like most mobile apps) rather than something you treat like a traditional desktop app, so I don't want to put too much focus on files.

>   - "open with" Saber when you click on those files,

I haven't figured out how to make a Flutter app a file handler for `.sbn` files, although I haven't looked very hard. I would like to at least assign the Saber icon to `.sbn` files even if they can't be opened (like the current behaviour on Windows).

> but that comes with the overhead of needing a database that keeps track of where are the files you opened to display the list within the app, unless you decide to only do that in the Saber directory of course.

I already keep a list of recently accessed files in SharedPreferences so this shouldn't be a problem: [lib/data/prefs.dart#L102](https://github.com/adil192/saber/blob/main/lib/data/prefs.dart#L102)

**Reply by quazar-omega** [reply]
> That's actually how it currently works!

Ah I apologize, I should have delved more into the codebase instead of making assumptions, anyway it's possible to achieve that on Android as well by hardcoding (I know, very bad, but it's the only way I found) `/storage/emulated/0/` + `Documents/Saber`, that would also require to ask for permission to access all files, the second bit could be a variable to to save in the preferences that can also be changed and reset by the user

> I'd like to keep this app mostly as a self contained tool

Yes, I agree, I made it sound like adding those features would change a lot, admittedly not knowing what was already there.


> I haven't figured out how to make a Flutter app a file handler for .sbn files

I found this library [open_filex](https://pub.dev/packages/open_filex), but it seems it doesn't support custom file associations, so yeah, that's a tough one

> I would like to at least assign the Saber icon to .sbn files

That would be pretty good!

> I already keep a list of recently accessed files

Oops, then it's already solved

**Reply by adil192** [reply]
> Ah I apologize, I should have delved more into the codebase instead of making assumptions, anyway it's possible to achieve that on Android as well by hardcoding (I know, very bad, but it's the only way I found) `/storage/emulated/0/` + `Documents/Saber`, that would also require to ask for permission to access all files, the second bit could be a variable to to save in the preferences that can also be changed and reset by the user

I don't expect you to know everything about the code! It should be relatively simple in theory to implement with [getExternalStorageDirectory()](https://pub.dev/documentation/path_provider/latest/path_provider/getExternalStorageDirectory.html) and a bit of refactoring, I just haven't got around to it yet.

**Reply by quazar-omega** [reply]
Thanks for being understanding!
Using `getExternalStorageDirectory()` to achieve that could be possible now that I think about it, I'll try to test that to see how that would work
